File: worker/frontdesk_worker/queue/create.py
"""Selection by QUEUE_PROVIDER only - no auto-detection, no other switch
(ADR-0004). Mirrors api/src/queue/create.ts, including its
createQueue/createIngestQueue split (ADR-0025 §1) around one shared
provider-branching helper.
"""

import asyncpg

from ..settings import Settings
from . import INGEST_QUEUE_NAME, QUEUE_NAME, Queue
from .pgmq import PgmqQueue
from .sqs import SqsQueue


def _create_queue_for(
    settings: Settings,
    pool: asyncpg.Pool,
    *,
    pgmq_name: str,
    sqs_queue_url: str | None,
    sqs_dlq_url: str | None,
) -> Queue:
    if settings.queue_provider == "pgmq":
        return PgmqQueue(pool, queue_name=pgmq_name)
    if settings.queue_provider == "sqs":
        if not sqs_queue_url or not sqs_dlq_url or not settings.aws_region:
            raise ValueError(
                "SQS_QUEUE_URL, SQS_DLQ_URL, and AWS_REGION are required when QUEUE_PROVIDER=sqs"
            )
        return SqsQueue(
            queue_url=sqs_queue_url,
            dlq_url=sqs_dlq_url,
            region=settings.aws_region,
            endpoint_url=settings.aws_endpoint_url,
        )
    raise ValueError(f"Unknown QUEUE_PROVIDER: {settings.queue_provider}")


def create_queue(settings: Settings, pool: asyncpg.Pool) -> Queue:
    return _create_queue_for(
        settings,
        pool,
        pgmq_name=QUEUE_NAME,
        sqs_queue_url=settings.sqs_queue_url,
        sqs_dlq_url=settings.sqs_dlq_url,
    )


def create_ingest_queue(settings: Settings, pool: asyncpg.Pool) -> Queue:
    return _create_queue_for(
        settings,
        pool,
        pgmq_name=INGEST_QUEUE_NAME,
        sqs_queue_url=settings.ingest_sqs_queue_url,
        sqs_dlq_url=settings.ingest_sqs_dlq_url,
    )

File: api/src/queue/create.ts
import type { Db } from "@frontdesk/db";
import type { Env } from "../env.js";
import {
  INGEST_QUEUE_NAME,
  QUEUE_NAME,
  type IngestMessage,
  type Queue,
  type TriageMessage,
} from "./index.js";
import { PostgresQueue } from "./pgmq.js";
import { SqsQueue } from "./sqs.js";

// Selection by QUEUE_PROVIDER only - no auto-detection, no other switch
// (ADR-0004). Shared by createQueue and createIngestQueue below: the two
// queues differ only in name and (under sqs) which URL pair backs them -
// everything else about "how to build a Queue for this provider" is
// identical, so it lives once here rather than twice.
function createQueueFor<T>(
  env: Env,
  db: Db,
  opts: { pgmqName: string; sqsQueueUrl: string | undefined; sqsDlqUrl: string | undefined },
): Queue<T> {
  if (env.QUEUE_PROVIDER === "pgmq") {
    return new PostgresQueue<T>(db, opts.pgmqName);
  }
  if (env.QUEUE_PROVIDER === "sqs") {
    if (!opts.sqsQueueUrl || !opts.sqsDlqUrl || !env.AWS_REGION) {
      throw new Error(
        "SQS_QUEUE_URL, SQS_DLQ_URL, and AWS_REGION are required when QUEUE_PROVIDER=sqs",
      );
    }
    return new SqsQueue<T>({
      queueUrl: opts.sqsQueueUrl,
      dlqUrl: opts.sqsDlqUrl,
      region: env.AWS_REGION,
      endpoint: env.AWS_ENDPOINT_URL,
    });
  }
  // Exhaustive per env.ts's z.enum(["pgmq", "sqs"]) - unreachable in
  // practice, but satisfies TypeScript in the general case.
  throw new Error(`Unknown QUEUE_PROVIDER: ${env.QUEUE_PROVIDER as string}`);
}

export function createQueue(env: Env, db: Db): Queue<TriageMessage> {
  return createQueueFor<TriageMessage>(env, db, {
    pgmqName: QUEUE_NAME,
    sqsQueueUrl: env.SQS_QUEUE_URL,
    sqsDlqUrl: env.SQS_DLQ_URL,
  });
}

// ADR-0025 §1: the ingestion trigger queue. Error text intentionally says
// "SQS_QUEUE_URL, SQS_DLQ_URL" (not the INGEST_ prefixed names) because
// that is the generic message createQueueFor always throws - loadEnv()'s
// own check (env.ts) is what names the INGEST_-prefixed keys specifically
// when they're the ones missing.
export function createIngestQueue(env: Env, db: Db): Queue<IngestMessage> {
  return createQueueFor<IngestMessage>(env, db, {
    pgmqName: INGEST_QUEUE_NAME,
    sqsQueueUrl: env.INGEST_SQS_QUEUE_URL,
    sqsDlqUrl: env.INGEST_SQS_DLQ_URL,
  });
}
